File: tessera-py/protocol/receipt.py
# VCR schema implementation v0.1
#
# canonical serialisation is binary: fields in spec order, each length-prefixed
# with 4-byte big-endian uint32. deterministic across any language/implementation.

from __future__ import annotations

import base64
import hashlib
import json
import struct
import time
from dataclasses import dataclass, field
from typing import Optional

from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)
from cryptography.hazmat.primitives.serialization import (
    Encoding,
    PublicFormat,
)

MAX_PROVENANCE_DEPTH = 256


class ReceiptValidationError(ValueError):
    """Raised when a receipt fails structural validation."""
    pass


@dataclass
class ParentRef:
    receipt_id: bytes       # bytes32
    receipt_hash: bytes     # bytes32 — SHA-256 of parent's canonical form
    relationship: str       # "input", "reference", or "aggregation"

    def canonical_bytes(self) -> bytes:
        return (
            _encode_field(self.receipt_id)
            + _encode_field(self.receipt_hash)
            + _encode_field(self.relationship.encode("utf-8"))
        )

    def to_dict(self) -> dict:
        return {
            "receipt_id": self.receipt_id.hex(),
            "receipt_hash": self.receipt_hash.hex(),
            "relationship": self.relationship,
        }

    @classmethod
    def from_dict(cls, d: dict) -> ParentRef:
        return cls(
            receipt_id=bytes.fromhex(d["receipt_id"]),
            receipt_hash=bytes.fromhex(d["receipt_hash"]),
            relationship=d["relationship"],
        )


@dataclass
class RoyaltyTerms:
    provider_royalty: int      # basis points (0-10000)
    parent_royalty: int = 0    # bps split equally among parents
    cascade: bool = False      # propagate up entire chain

    def validate(self) -> None:
        if not (0 <= self.provider_royalty <= 10000):
            raise ReceiptValidationError(
                f"provider_royalty must be 0–10000 bps, got {self.provider_royalty}"
            )
        if not (0 <= self.parent_royalty <= 10000):
            raise ReceiptValidationError(
                f"parent_royalty must be 0–10000 bps, got {self.parent_royalty}"
            )
        if self.provider_royalty + self.parent_royalty > 10000:
            raise ReceiptValidationError(
                f"royalty sum exceeds 10000 bps: {self.provider_royalty} + {self.parent_royalty}"
                f" = {self.provider_royalty + self.parent_royalty}"
            )

    def canonical_bytes(self) -> bytes:
        return (
            _encode_field(struct.pack(">H", self.provider_royalty))
            + _encode_field(struct.pack(">H", self.parent_royalty))
            + _encode_field(struct.pack(">?", self.cascade))
        )

    def to_dict(self) -> dict:
        return {
            "provider_royalty": self.provider_royalty,
            "parent_royalty": self.parent_royalty,
            "cascade": self.cascade,
        }

    @classmethod
    def from_dict(cls, d: dict) -> RoyaltyTerms:
        return cls(**d)


@dataclass
class Extension:
    type: str
    data: bytes

    def canonical_bytes(self) -> bytes:
        return (
            _encode_field(self.type.encode("utf-8"))
            + _encode_field(self.data)
        )

    def to_dict(self) -> dict:
        return {"type": self.type, "data": base64.b64encode(self.data).decode()}

    @classmethod
    def from_dict(cls, d: dict) -> Extension:
        return cls(type=d["type"], data=base64.b64decode(d["data"]))


@dataclass
class Receipt:
    # identity
    schema_version: int = 1

    # computation
    model_id: bytes = b""
    verification_key_id: bytes = b""
    input_hash: bytes = b""
    output_hash: bytes = b""
    proof: bytes = b""
    public_inputs: bytes = b""
    proving_backend: str = "ezkl-halo2"
    timestamp: int = field(default_factory=lambda: int(time.time()))

    # provenance
    parent_receipts: list[ParentRef] = field(default_factory=list)
    provenance_depth: int = 0

    # economics
    provider: bytes = b""
    original_price: int = 0
    currency: str = "USD-cents"
    royalty_terms: RoyaltyTerms = field(default_factory=lambda: RoyaltyTerms(provider_royalty=500))
    transfer_count: int = 0  # mutable — excluded from canonical hash, tracked by transfer ledger

    # integrity
    # receipt_id and receipt_hash are the same derived value:
    # SHA-256(canonical(all fields above, excluding transfer_count and signature)).
    # never stored, always computed.
    signature: bytes = b""
    signature_scheme: str = "ed25519"

    # extensions
    extensions: list[Extension] = field(default_factory=list)

    # not part of the VCR — travels alongside, excluded from hashes/signatures
    output_data: Optional[bytes] = None

    def validate(self) -> None:
        """Check structural validity. Called before signing and at ingestion."""
        self.royalty_terms.validate()

        # duplicate parent receipt_ids
        seen = set()
        for p in self.parent_receipts:
            pid = bytes(p.receipt_id)
            if pid in seen:
                raise ReceiptValidationError(
                    f"duplicate parent receipt_id: {pid.hex()}"
                )
            seen.add(pid)

        # provenance depth cap
        if self.provenance_depth > MAX_PROVENANCE_DEPTH:
            raise ReceiptValidationError(
                f"provenance_depth {self.provenance_depth} exceeds maximum {MAX_PROVENANCE_DEPTH}"
            )

    def canonical_bytes(self) -> bytes:
        parts = []

        parts.append(_encode_field(struct.pack(">H", self.schema_version)))

        parts.append(_encode_field(self.model_id))
        parts.append(_encode_field(self.verification_key_id))
        parts.append(_encode_field(self.input_hash))
        parts.append(_encode_field(self.output_hash))
        parts.append(_encode_field(self.proof))
        parts.append(_encode_field(self.public_inputs))
        parts.append(_encode_field(self.proving_backend.encode("utf-8")))
        parts.append(_encode_field(struct.pack(">Q", self.timestamp)))

        parts.append(_encode_list(self.parent_receipts))
        parts.append(_encode_field(struct.pack(">H", self.provenance_depth)))

        parts.append(_encode_field(self.provider))
        parts.append(_encode_field(struct.pack(">Q", self.original_price)))
        parts.append(_encode_field(self.currency.encode("utf-8")))
        parts.append(self.royalty_terms.canonical_bytes())
        # transfer_count is excluded from canonical hash — it's mutable state
        # tracked by the transfer ledger, not part of the immutable receipt

        parts.append(_encode_field(self.signature_scheme.encode("utf-8")))

        parts.append(_encode_list(self.extensions))

        return b"".join(parts)

    @property
    def receipt_id(self) -> bytes:
        return hashlib.sha256(self.canonical_bytes()).digest()

    @property
    def receipt_hash(self) -> bytes:
        return self.receipt_id

    def receipt_id_hex(self) -> str:
        return self.receipt_id.hex()

    def receipt_hash_hex(self) -> str:
        return self.receipt_hash.hex()

    def sign(self, private_key: Ed25519PrivateKey) -> None:
        self.provider = private_key.public_key().public_bytes(
            Encoding.Raw, PublicFormat.Raw
        )
        self.validate()
        self.signature = private_key.sign(self.receipt_hash)

    def verify_signature(self) -> bool:
        if not self.signature or not self.provider:
            return False
        try:
            pubkey = Ed25519PublicKey.from_public_bytes(self.provider)
            pubkey.verify(self.signature, self.receipt_hash)
            return True
        except Exception:
            return False

    def verify_integrity(self) -> bool:
        # receipt_hash is a property that always recomputes from fields,
        # so integrity is structural for in-memory receipts.
        return True

    def as_parent_ref(self, relationship: str = "input") -> ParentRef:
        return ParentRef(
            receipt_id=self.receipt_id,
            receipt_hash=self.receipt_hash,
            relationship=relationship,
        )

    def compute_provenance_depth(self) -> int:
        if not self.parent_receipts:
            return 0
        return self.provenance_depth

    def to_json(self) -> str:
        d = {
            "schema_version": self.schema_version,
            "model_id": self.model_id.hex(),
            "verification_key_id": self.verification_key_id.hex(),
            "input_hash": self.input_hash.hex(),
            "output_hash": self.output_hash.hex(),
            "proof": base64.b64encode(self.proof).decode(),
            "public_inputs": base64.b64encode(self.public_inputs).decode(),
            "proving_backend": self.proving_backend,
            "timestamp": self.timestamp,
            "parent_receipts": [p.to_dict() for p in self.parent_receipts],
            "provenance_depth": self.provenance_depth,
            "provider": self.provider.hex(),
            "original_price": self.original_price,
            "currency": self.currency,
            "royalty_terms": self.royalty_terms.to_dict(),
            "transfer_count": self.transfer_count,
            "signature": base64.b64encode(self.signature).decode(),
            "signature_scheme": self.signature_scheme,
            "extensions": [e.to_dict() for e in self.extensions],
        }
        if self.output_data is not None:
            d["output_data"] = base64.b64encode(self.output_data).decode()
        return json.dumps(d, sort_keys=True, separators=(",", ":"))

    @classmethod
    def from_json(cls, raw: str) -> Receipt:
        d = json.loads(raw)
        r = cls(
            schema_version=d["schema_version"],
            model_id=bytes.fromhex(d["model_id"]),
            verification_key_id=bytes.fromhex(d["verification_key_id"]),
            input_hash=bytes.fromhex(d["input_hash"]),
            output_hash=bytes.fromhex(d["output_hash"]),
            proof=base64.b64decode(d["proof"]),
            public_inputs=base64.b64decode(d["public_inputs"]),
            proving_backend=d["proving_backend"],
            timestamp=d["timestamp"],
            parent_receipts=[ParentRef.from_dict(p) for p in d["parent_receipts"]],
            provenance_depth=d["provenance_depth"],
            provider=bytes.fromhex(d["provider"]),
            original_price=d["original_price"],
            currency=d["currency"],
            royalty_terms=RoyaltyTerms.from_dict(d["royalty_terms"]),
            transfer_count=d["transfer_count"],
            signature=base64.b64decode(d.get("signature", "")),
            signature_scheme=d.get("signature_scheme", "ed25519"),
            extensions=[Extension.from_dict(e) for e in d.get("extensions", [])],
        )
        if "output_data" in d:
            r.output_data = base64.b64decode(d["output_data"])
        return r


def _encode_field(data: bytes) -> bytes:
    return struct.pack(">I", len(data)) + data


def _encode_list(items: list) -> bytes:
    parts = [struct.pack(">I", len(items))]
    for item in items:
        parts.append(item.canonical_bytes())
    return b"".join(parts)

File: tessera-rust/tessera/src/receipt.rs
//! VCR receipt schema and canonical serialisation.
//!
//! Implements Section 2–5 of VCR-SPEC.md.

use ed25519_dalek::{Signer, SigningKey, Verifier, VerifyingKey};
use sha2::{Digest, Sha256};

// ── Supporting structures (Section 3) ──────────────────────────

#[derive(Clone, Debug)]
pub struct ParentRef {
    pub receipt_id: Vec<u8>,   // bytes32
    pub receipt_hash: Vec<u8>, // bytes32
    pub relationship: String,  // "input", "reference", "aggregation"
}

impl ParentRef {
    /// Canonical bytes: 3 LEN-prefixed sub-fields (Section 4.3).
    pub fn canonical_bytes(&self) -> Vec<u8> {
        let mut out = Vec::new();
        encode_field(&mut out, &self.receipt_id);
        encode_field(&mut out, &self.receipt_hash);
        encode_field(&mut out, self.relationship.as_bytes());
        out
    }
}

#[derive(Clone, Debug)]
pub struct RoyaltyTerms {
    pub provider_royalty: u16, // basis points 0–10000
    pub parent_royalty: u16,
    pub cascade: bool,
}

impl RoyaltyTerms {
    /// Canonical bytes: 3 LEN-prefixed sub-fields, no outer wrapper (Section 4.3).
    pub fn canonical_bytes(&self) -> Vec<u8> {
        let mut out = Vec::new();
        encode_field(&mut out, &self.provider_royalty.to_be_bytes());
        encode_field(&mut out, &self.parent_royalty.to_be_bytes());
        encode_field(&mut out, &[if self.cascade { 0x01 } else { 0x00 }]);
        out
    }
}

#[derive(Clone, Debug)]
pub struct Extension {
    pub ext_type: String,
    pub data: Vec<u8>,
}

impl Extension {
    pub fn canonical_bytes(&self) -> Vec<u8> {
        let mut out = Vec::new();
        encode_field(&mut out, self.ext_type.as_bytes());
        encode_field(&mut out, &self.data);
        out
    }
}

// ── Receipt (Section 2) ────────────────────────────────────────

#[derive(Clone, Debug)]
pub struct Receipt {
    // Identity
    pub schema_version: u16,

    // Computation
    pub model_id: Vec<u8>,
    pub verification_key_id: Vec<u8>,
    pub input_hash: Vec<u8>,
    pub output_hash: Vec<u8>,
    pub proof: Vec<u8>,
    pub public_inputs: Vec<u8>,
    pub proving_backend: String,
    pub timestamp: u64,

    // Provenance
    pub parent_receipts: Vec<ParentRef>,
    pub provenance_depth: u16,

    // Economics
    pub provider: Vec<u8>,
    pub original_price: u64,
    pub currency: String,
    pub royalty_terms: RoyaltyTerms,
    pub transfer_count: u16, // excluded from canonical serialisation

    // Integrity
    pub signature: Vec<u8>,  // excluded from canonical serialisation
    pub signature_scheme: String,
    pub extensions: Vec<Extension>,
}

impl Receipt {
    /// Canonical binary serialisation (Section 4.4).
    ///
    /// 17 fields in exact normative order. transfer_count and signature excluded.
    pub fn canonical_bytes(&self) -> Vec<u8> {
        let mut out = Vec::new();

        //  1. schema_version
        encode_field(&mut out, &self.schema_version.to_be_bytes());
        //  2. model_id
        encode_field(&mut out, &self.model_id);
        //  3. verification_key_id
        encode_field(&mut out, &self.verification_key_id);
        //  4. input_hash
        encode_field(&mut out, &self.input_hash);
        //  5. output_hash
        encode_field(&mut out, &self.output_hash);
        //  6. proof
        encode_field(&mut out, &self.proof);
        //  7. public_inputs
        encode_field(&mut out, &self.public_inputs);
        //  8. proving_backend
        encode_field(&mut out, self.proving_backend.as_bytes());
        //  9. timestamp
        encode_field(&mut out, &self.timestamp.to_be_bytes());
        // 10. parent_receipts (LIST)
        encode_list(&mut out, &self.parent_receipts);
        // 11. provenance_depth
        encode_field(&mut out, &self.provenance_depth.to_be_bytes());
        // 12. provider
        encode_field(&mut out, &self.provider);
        // 13. original_price
        encode_field(&mut out, &self.original_price.to_be_bytes());
        // 14. currency
        encode_field(&mut out, self.currency.as_bytes());
        // 15. royalty_terms (nested, no outer wrapper)
        out.extend_from_slice(&self.royalty_terms.canonical_bytes());
        // 16. signature_scheme
        encode_field(&mut out, self.signature_scheme.as_bytes());
        // 17. extensions (LIST)
        encode_ext_list(&mut out, &self.extensions);

        out
    }

    /// receipt_id = SHA-256(canonical_bytes) (Section 5.1).
    pub fn receipt_id(&self) -> Vec<u8> {
        let mut hasher = Sha256::new();
        hasher.update(&self.canonical_bytes());
        hasher.finalize().to_vec()
    }

    /// receipt_hash is an alias for receipt_id (Section 5.1).
    pub fn receipt_hash(&self) -> Vec<u8> {
        self.receipt_id()
    }

    /// Sign the receipt with an Ed25519 private key (Section 5.3).
    ///
    /// Sets provider to the corresponding public key and computes signature.
    pub fn sign(&mut self, signing_key: &SigningKey) {
        self.provider = signing_key.verifying_key().to_bytes().to_vec();
        let hash = self.receipt_hash();
        self.signature = signing_key.sign(&hash).to_bytes().to_vec();
    }

    /// Verify the signature against the provider public key (Section 6.2).
    pub fn verify_signature(&self) -> bool {
        if self.signature.len() != 64 || self.provider.len() != 32 {
            return false;
        }
        let Ok(vk_bytes): Result<[u8; 32], _> = self.provider.clone().try_into() else {
            return false;
        };
        let Ok(verifying_key) = VerifyingKey::from_bytes(&vk_bytes) else {
            return false;
        };
        let Ok(sig_bytes): Result<[u8; 64], _> = self.signature.clone().try_into() else {
            return false;
        };
        let signature = ed25519_dalek::Signature::from_bytes(&sig_bytes);
        let hash = self.receipt_hash();
        verifying_key.verify(&hash, &signature).is_ok()
    }
}

// ── Encoding helpers (Section 4.1–4.2) ─────────────────────────

/// LEN(x) = BE32(len(x)) || x
fn encode_field(out: &mut Vec<u8>, data: &[u8]) {
    out.extend_from_slice(&(data.len() as u32).to_be_bytes());
    out.extend_from_slice(data);
}

/// LIST = BE32(count) || item_0.canonical_bytes() || ...
fn encode_list(out: &mut Vec<u8>, items: &[ParentRef]) {
    out.extend_from_slice(&(items.len() as u32).to_be_bytes());
    for item in items {
        out.extend_from_slice(&item.canonical_bytes());
    }
}

fn encode_ext_list(out: &mut Vec<u8>, items: &[Extension]) {
    out.extend_from_slice(&(items.len() as u32).to_be_bytes());
    for item in items {
        out.extend_from_slice(&item.canonical_bytes());
    }
}
